flowmachine/flowmachine/core/cache.py

```diff
@@ -0,0 +1,384 @@
+
+    logger.info(
+        f"{'Would' if dry_run else 'Will'} remove cache record for {obj_to_remove.md5} of type {obj_to_remove.__class__}"
+    )
+    logger.info(
+        f"Table {obj_to_remove.table_name} ({obj_size} bytes) {'would' if dry_run else 'will'} be removed."
+    )
+
+    if not dry_run:
+        obj_to_remove.invalidate_db_cache(cascade=False, drop=True)
+    return obj_to_remove, obj_size
+
+
+def shrink_below_size(
+    connection: "Connection", size_threshold: int = None, dry_run: bool = False
+) -> "Query":
+    """
+    Remove queries from the cache until it is below a specified size threshold.
+
+    Parameters
+    ----------
+    connection : "Connection"
+    size_threshold : int, default None
+        Optionally override the maximum cache size set in flowdb.
+    dry_run : bool, default False
+        Set to true to just report the objects that would be removed and not remove them
+
+    Returns
+    -------
+    list of "Query"
+        List of the queries that were removed
+    """
+    initial_cache_size = size_of_cache(connection)
+    removed = []
+    logger.info(
+        f"Shrinking cache from {initial_cache_size} to below {size_threshold}{'(dry run)' if dry_run else ''}."
+    )
+
+    if dry_run:
+        cached_queries = iter(get_cached_queries_by_score(connection))
+        shrink = lambda x: cached_queries.__next__()
+    else:
+        shrink = shrink_one
+
+    if size_threshold is None:
+        size_threshold = get_max_size_of_cache(connection)
+
+    while initial_cache_size > size_threshold:
+        obj_removed, cache_reduction = shrink(connection)
+        removed.append(obj_removed)
+        initial_cache_size -= cache_reduction
+    logger.info(
+        f"New cache size {'would' if dry_run else 'will'} be {initial_cache_size}."
+    )
+    return removed
+
+
+def size_of_table(connection: "Connection", table_name: str, table_schema: str) -> int:
+    """
+    Get the size on disk in bytes of a table in the database.
+
+    Parameters
+    ----------
+    connection : "Connection"
+    table_name : str
+        Name of table to get size of
+    table_schema : str
+        Schema of the table
+
+    Returns
+    -------
+    int
+        Number of bytes on disk this table uses in total
+
+    """
+    sql = f"SELECT table_size('{table_name}', '{table_schema}')"
+    try:
+        return int(connection.fetch(sql)[0][0])
+    except IndexError:
+        raise ValueError(
+            f"Table '{table_schema}.{table_name}' does not exist on this connection."
+        )
+
+
+def size_of_cache(connection: "Connection") -> int:
+    """
+    Get the total size in bytes of all cache tables.
+
+    Parameters
+    ----------
+    connection : "Connection"
+
+    Returns
+    -------
+    int
+        Number of bytes in total used by cache tables
+
+    """
+    sql = """SELECT sum(table_size(tablename, schema)) as total_bytes 
+        FROM cache.cached  
+        WHERE NOT cached.class='Table'"""
```

Comment:
It occurs to me that almost all of the queries have an explicit exclusion clause of the form `WHERE NOT cached.class='Table'` (or similar). Maybe not for this PR, but is it worth storing `Table` objects somewhere else so that we don't need to continually treat them as an exception?